Lidar test fix for #1712

amcastro-tri opened the pull request:
This is a fix for the bug reported in #1712.

The discussion in #1712 briefly explains the problem. Here I'll summarize the findings:
- Bullet is doing the final raycast with [btSubsimplexConvexCast::calcTimeOfImpact](https://github.com/bulletphysics/bullet3/blob/master/src/BulletCollision/NarrowPhaseCollision/btSubSimplexConvexCast.cpp#L38).
- This is an iterative procedure to find the ray/geometry intersection.
- There is a hardcoded tolerance to be [epsilon = 0.0001](https://github.com/bulletphysics/bullet3/blob/master/src/BulletCollision/NarrowPhaseCollision/btSubSimplexConvexCast.cpp#L76)).
- For the "failed" cases what actually happens is that the method returns but the convergence is quite bad and only converging within `0.0001`. Actually the root square of this (0.01) since the value is being compared with a squared distance (this explains the error of magnitude in the errors observed in #1712).

The solution is to change the default algorithm (the sub simplex convex cast) with a GJK convex cast. This is done by setting `ClosestRayResultCallback` to have the `btTriangleRaycastCallback::kF_UseGjkConvexCastRaytest` bit active as done [here](https://github.com/amcastro-tri/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/collision/bullet_model.cc#L611).

The lidar test is run with a much smaller angle span to highlight the area where the problem happens as in #1712. This span is for angles in `(-0.1; 0.1)` radians. The figure below shows in red the distances returned by the lidar as computed with the default sub-simplex convex cast algorithm. The circles show the lidar returns computed with the fix. The solid black line corresponds to the analytical value. 
Notice that now the tolerance for the unit test is reduced to 1.0e-6 and the test successfully passes (before we were passing it with a tolerance of 0.01 to skip this error).

![lidarcloseup_fix2](https://cloud.githubusercontent.com/assets/17601461/15265541/d924f890-1954-11e6-914d-b36e353cabc6.jpg)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2354) &emsp; Multiple assignees:&emsp;<img alt="@RussTedrake" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/6442292?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/RussTedrake">RussTedrake</a>,&emsp;<img alt="@amcastro-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17601461?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/amcastro-tri">amcastro-tri</a>,&emsp;<img alt="@jwnimmer-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596505?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/jwnimmer-tri">jwnimmer-tri</a>
<!-- Reviewable:end -->

jwnimmer-tri commented:
Wow, looks great!  Small PR, builds atop @liangfok's recent improvements, reduces magic constants, removes magic `auto`, new comments. :+1:

amcastro-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 613 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzikBXz2by-mHu1Mr:-KHgzikBXz2by-mHu1Ms:2010057998) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/collision/bullet_model.cc#L613)):*
> ```cpp
>     ray_callback.m_flags |=
>         btTriangleRaycastCallback::kF_UseGjkConvexCastRaytest;
> ```

This is the actual change that fixes the problem.

---

*[drake/systems/plants/collision/bullet_model.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzmYokP7DAcS9Cj2T:-KHgzmYpPf330DtZvrcx:-1364997454) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/collision/bullet_model.h#L11)):*
> ```objc
> #include "BulletCollision/NarrowPhaseCollision/btPointCollector.h"
> #include "BulletCollision/NarrowPhaseCollision/btRaycastCallback.h"
> ```

Header needed for the definition of `btTriangleRaycastCallback::kF_UseGjkConvexCastRaytest`.

---

*[drake/systems/plants/test/lidarTest.cpp, line 74 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzxnfogqqBw-gfTKl:-KHgzxnfogqqBw-gfTKm:1705797923) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/test/lidarTest.cpp#L74)):*
> ```cpp
>   const double max_yaw = lidar_sensor->max_yaw();
>   const double max_range = lidar_sensor->max_range();
>   const double tol = 1.0e-6;
> ```

The tolerance can now be reduced from `0.01` to `1.0e-6`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@RussTedrake, +@sherm1 

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Wow, looks great!  Small PR, builds atop @liangfok's recent improvements, reduces magic constants, removes magic `auto`, new comments. :+1:</details>


---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHh-CBB7LLq0rXrHXEz:-213039246)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
haha, yeah, amazing what a single line of code can do.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> +@RussTedrake, +@sherm1 </details>


---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHh-KXFJ6i0tGkMof5o:-341896977)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Great work!

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> haha, yeah, amazing what a single line of code can do.</details>


---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.cpp, line 612 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh0iHQML46sMcmRVjJ:-KHh0iHRSxZmxFBrJXeF:-1801852285) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.cpp#L612)):*
> ```cpp
> 
>   // Computes the origin of the rays (at [0,0,0] in the frame of the sensor) in
>   // the world frame.
> ```

Consider splitting the above comment into two separate sentences:

```
// Computes the origin of the rays in the world frame. The origin of
// of the rays in the sensor's frame is [0,0,0].
```

---

*[drake/systems/plants/RigidBodySystem.h, line 546 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh1GI23NWuN6zlFp1s:-KHh1GI23NWuN6zlFp1t:-1279741188) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L546)):*
> ```objc
>   void CheckValidConfiguration();
> 
>   // Depth sensor will ray cast a ray with its start point at (0,0,0) (in the
> ```

Consider:

"Depth" -> "The depth"

---

*[drake/systems/plants/RigidBodySystem.h, line 546 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh1SaVG9aE7mpyupWh:-KHh1SaVG9aE7mpyupWi:67959878) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L546)):*
> ```objc
>   void CheckValidConfiguration();
> 
>   // Depth sensor will ray cast a ray with its start point at (0,0,0) (in the
> ```

Question: Is there a way to reword the phrase "ray cast a ray"? It's awkward having two "ray" words so close to each other.

---

*[drake/systems/plants/RigidBodySystem.h, line 551 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh2iLgRnlt5y_7ty_i:-KHh2iLgRnlt5y_7ty_j:603752851) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L551)):*
> ```objc
>   // moment of construction. raycast_endpoints is only computed once at
>   // construction since the end points are constant in the frame of the sensor.
>   // The end points are coputed by scanning in the yaw (pitch) direction
> ```

Typo: "coputed"

---

*[drake/systems/plants/RigidBodySystem.h, line 551 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh2szwn6N6rVY4YL5M:-KHh2szwn6N6rVY4YL5N:1550855803) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L551)):*
> ```objc
>   // moment of construction. raycast_endpoints is only computed once at
>   // construction since the end points are constant in the frame of the sensor.
>   // The end points are coputed by scanning in the yaw (pitch) direction
> ```

Consider: "yaw (pitch)" -> "yaw or pitch"

---

*[drake/systems/plants/RigidBodySystem.h, line 552 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh2xslAKSmG7aHLQh7:-KHh2xslAKSmG7aHLQh8:-1531154689) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L552)):*
> ```objc
>   // construction since the end points are constant in the frame of the sensor.
>   // The end points are coputed by scanning in the yaw (pitch) direction
>   // discretizing the yaw (pitch) range in num_pixel_cols (num_pixel_rows).
> ```

Consider: "yaw (pitch)" -> "yaw or pitch".
Consider: "num_pixel_cols (num_pixel_rows)" -> "num_pixel_cols or num_pixel_rows"

---

*[drake/systems/plants/RigidBodySystem.h, line 554 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh3F0CRfL6WiSi_x3x:-KHh3F0CRfL6WiSi_x3y:1305705727) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L554)):*
> ```objc
>   // discretizing the yaw (pitch) range in num_pixel_cols (num_pixel_rows).
>   // The final 3D end point then corresponds to a ray that starts at zero, with
>   // length max_range, at the specific yaw (pitch) angle.
> ```

Consider: "yaw (pitch)" - "yaw or pitch"

---

*[drake/systems/plants/collision/bullet_model.cc, line 613 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzikBXz2by-mHu1Mr:-KHh3ONeydMzYAQ25bSp:-215697932) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/collision/bullet_model.cc#L613)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This is the actual change that fixes the problem.</details>
Question: Nice. Can you add a comment about potential performance implications of this change and alternative algorithm choices? Maybe just including a link to documentation in bullet that lists the various choices is sufficient.

---

*[drake/systems/plants/test/lidarTest.cpp, line 74 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzxnfogqqBw-gfTKl:-KHh4ODELh9xHf2_7IgD:746934339) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/test/lidarTest.cpp#L74)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> The tolerance can now be reduced from `0.01` to `1.0e-6`</details>
Awesome!

---

*[drake/systems/plants/test/lidarTest.cpp, line 79 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh4SFXWZ9ReYVTxwRU:-KHh4SFXWZ9ReYVTxwRV:1293218157) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/test/lidarTest.cpp#L79)):*
> ```cpp
>     double theta = min_yaw + (max_yaw - min_yaw) * i / (distances.size() - 1);
> 
>     // We've implicitly hard-coded the box geometry from the sdf for the
> ```

Defect: "sdf" -> "SDF"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHh4_aqUqQfDtvFh2Wn:-567554664)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.cpp, line 612 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh0iHQML46sMcmRVjJ:-KHjS63xbXPfBUHrQ-TB:51486461) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.cpp#L612)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Consider splitting the above comment into two separate sentences:
> 
> ```
> // Computes the origin of the rays in the world frame. The origin of
> // of the rays in the sensor's frame is [0,0,0].
> ```</details>
Considered

---

*[drake/systems/plants/RigidBodySystem.h, line 546 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh1GI23NWuN6zlFp1s:-KHjSTEx3olU1Lv6O0i3:-1941978403) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L546)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Consider:
> 
> "Depth" -> "The depth"</details>
considered

---

*[drake/systems/plants/RigidBodySystem.h, line 546 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh1SaVG9aE7mpyupWh:-KHjSf1sMnjgV6QSQHWf:-1274397607) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L546)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Question: Is there a way to reword the phrase "ray cast a ray"? It's awkward having two "ray" words so close to each other.</details>
Agree. I removed the first "ray" and I think it sounds better. Does it?

---

*[drake/systems/plants/RigidBodySystem.h, line 551 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh2iLgRnlt5y_7ty_i:-KHjT5DXrRSGw3chJnYM:422879044) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L551)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Typo: "coputed"</details>
Good catch!

---

*[drake/systems/plants/RigidBodySystem.h, line 551 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh2szwn6N6rVY4YL5M:-KHjT6mcC9UOIF1BkkRG:-220270981) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L551)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Consider: "yaw (pitch)" -> "yaw or pitch"</details>
I like how it reads if you ignore the second option in parentheses because right now is the most common case (or only case). I will stick with the parentheses if you don't mind. 

---

*[drake/systems/plants/collision/bullet_model.cc, line 613 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHgzikBXz2by-mHu1Mr:-KHjrOJ8mQ7JDX5KFtG0:-798566094) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/collision/bullet_model.cc#L613)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Question: Nice. Can you add a comment about potential performance implications of this change and alternative algorithm choices? Maybe just including a link to documentation in bullet that lists the various choices is sufficient.</details>
I added all I currently know about this in a developers comment.

---

*[drake/systems/plants/test/lidarTest.cpp, line 79 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh4SFXWZ9ReYVTxwRU:-KHjTp5R3__8xShilQW9:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/test/lidarTest.cpp#L79)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Defect: "sdf" -> "SDF"</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Found a few typos.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Great work!</details>


---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.h, line 546 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHh1SaVG9aE7mpyupWh:-KHkFGucYR2P8fTzIzWV:-1809122099) ([raw file](https://github.com/robotlocomotion/drake/blob/2acf93150d1cc5ca336414d853220eb9ca855fee/drake/systems/plants/RigidBodySystem.h#L546)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Agree. I removed the first "ray" and I think it sounds better. Does it?</details>
LGTM. Thanks!

---

*[drake/systems/plants/collision/bullet_model.cc, line 624 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFg-LwkZGmxVspcmT:-KHkFg-LwkZGmxVspcmU:-2126348706) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L624)):*
> ```cpp
>     // - SubSimplexConvexCastRaytest is the default.
>     // - SubSimplexConvexCastRaytest uses an approximate but faster ray versus
>     //   convex intersection algorithm.
> ```

"convex" -> "a convex"

Or, should it be "the convex"?

---

*[drake/systems/plants/collision/bullet_model.cc, line 626 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFoRlD4jWuhni0-Hp:-KHkFoRlD4jWuhni0-Hq:-2033162900) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L626)):*
> ```cpp
>     //   convex intersection algorithm.
>     // We now know that SubSimplexConvexCastRaytest is an iterative algorithm
>     // with a quite large hardcoded tolerance for convergence, see discussions
> ```

"a quite" -> "a"

---

*[drake/systems/plants/collision/bullet_model.cc, line 631 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkG1vL-HPS-7-2wj14:-KHkG1vL-HPS-7-2wj15:423965087) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L631)):*
> ```cpp
>     // When using SubSimplexConvexCastRaytest the ray test finally gets resolved
>     // with the call to btSubsimplexConvexCast::calcTimeOfImpact() (this is the
>     // iterative method with the hardcoded tolerance.
> ```

Missing close parenthesis.

---

*[drake/systems/plants/collision/bullet_model.cc, line 632 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkG5bg1zfk4hhJmVh3:-KHkG5bg1zfk4hhJmVh4:1915250566) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L632)):*
> ```cpp
>     // with the call to btSubsimplexConvexCast::calcTimeOfImpact() (this is the
>     // iterative method with the hardcoded tolerance.
>     // A nice discussion (an probably the only one out there) is referenced at
> ```

"an" -> "and"

---

*[drake/systems/plants/collision/bullet_model.cc, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGA6Z_qC0LmH5H3VK:-KHkGA6Z_qC0LmH5H3VL:2036502564) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L634)):*
> ```cpp
>     // A nice discussion (an probably the only one out there) is referenced at
>     // the top of the file, quote:
>     // Typically the conservative advancement reaches solution in a few
> ```

I know this is probably  quote from bullet's code, but: "solution" -> "the solution"

---

*[drake/systems/plants/collision/bullet_model.cc, line 636 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGIPFsVzPZZgNtZWz:-KHkGIPFsVzPZZgNtZX-:2145256118) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L636)):*
> ```cpp
>     // Typically the conservative advancement reaches solution in a few
>     // iterations, clip it to 32 for degenerate cases. See discussion about this
>     // here http://continuousphysics.com/Bullet/phpBB2/viewtopic.php?t=565
> ```

"here" -> "here:"

---

*[drake/systems/plants/collision/bullet_model.h, line 126 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGf4ULftiDa95xoL7:-KHkGf4ULftiDa95xoL8:-1748787498) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.h#L126)):*
> ```objc
>   // Possible ray casting algorithms for Bullet.
>   enum RayCastAlgorithm {
>     SubSimplexConvexCast,
> ```

Question: Should these enums start with "k" to indicate they are constants?

---

*[drake/systems/plants/collision/bullet_model.h, line 160 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGaVbG7qzSIgtNigi:-KHkGaVbG7qzSIgtNigj:-481152905) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.h#L160)):*
> ```objc
>   BulletCollisionWorldWrapper bullet_world_no_margin_;
> 
>   // The Bullet algorithm to bse used for ray casting.
> ```

"bse" -> "be"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHkGvcmvgVDudwDfm_p:-2059045687)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 624 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFg-LwkZGmxVspcmT:-KHog9aSlnVWwOoPbkbA:-528741797) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L624)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "convex" -> "a convex"
> 
> Or, should it be "the convex"?</details>
I think the comment is ok. I understand it as referring not to a particular convex object but to the class of objects that are convex. In any case, I am quoting here so my preference is to leave it as it is in the original. I'm open though

---

*[drake/systems/plants/collision/bullet_model.cc, line 626 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFoRlD4jWuhni0-Hp:-KHogdGiaLO-LnnXyr32:755766863) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L626)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "a quite" -> "a"</details>
I changed to "a very large". Is the "quite" the problem? I am trying to stress that this hard coded tolerance is really ridiculously large (1 cm. Imagine what would happen when simulating Valkyrie fingers!). My English might be failing though :-)

---

*[drake/systems/plants/collision/bullet_model.cc, line 631 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkG1vL-HPS-7-2wj14:-KHohGvgzhGLQVbQPWcK:-2097206076) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L631)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Missing close parenthesis.</details>
nice catch!

---

*[drake/systems/plants/collision/bullet_model.cc, line 632 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkG5bg1zfk4hhJmVh3:-KHohQ2V4M6wTdRTztJD:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L632)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "an" -> "and"</details>
Done.

---

*[drake/systems/plants/collision/bullet_model.cc, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGA6Z_qC0LmH5H3VK:-KHohSfzMEX3GTY_nZij:250244091) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L634)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I know this is probably  quote from bullet's code, but: "solution" -> "the solution"</details>
Agree. But yes, I am trying to quote as-is from that Bullet comment so people can find it in their source. Open if you really don't like it. 

---

*[drake/systems/plants/collision/bullet_model.cc, line 636 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGIPFsVzPZZgNtZWz:-KHohee_JuTWstCIJMu1:-1906752578) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L636)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "here" -> "here:"</details>
as above, it's part of the quote.

---

*[drake/systems/plants/collision/bullet_model.h, line 126 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGf4ULftiDa95xoL7:-KHoi8YxGoKAh7bM8AuN:-1649137172) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.h#L126)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Question: Should these enums start with "k" to indicate they are constants?</details>
Great! I din't know about this, thank you. yes, Google prefers the "k" in front of specific enums while the enum itself follows the convention for types (it is a type).

---

*[drake/systems/plants/collision/bullet_model.h, line 160 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkGaVbG7qzSIgtNigi:-KHoinky_4KuXLMXiu2x:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.h#L160)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "bse" -> "be"</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks @liangfok for finding those typos!

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Found a few typos.</details>


---

Review status: 3 of 5 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHoiqM1U9vEgpXSKa8c:-830637504)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Beautiful! Great investigation work and a nice fix. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>. 

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Thanks @liangfok for finding those typos!</details>


---

Reviewed 3 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 615 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHprZfbhVvWbu08jeOr:-KHprZfbhVvWbu08jeOs:-941806159) ([raw file](https://github.com/robotlocomotion/drake/blob/eaa05a61a2c9587d78dc60e1143b582334f2c2f2/drake/systems/plants/collision/bullet_model.cc#L615)):*
> ```cpp
>     // RayResultCallback::m_flags.
>     // This is demonstrated in the RaytestDemo
>     // (bullet3/examplex/Raycast/RaytestDemo.cpp) part of Bullet's
> ```

examples?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHps_GJWSwdIcnMglx-:-465918299)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Very nice.  Thanks!  To be clear, judging by the name this will only work for convex geometries, correct?

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Beautiful! Great investigation work and a nice fix. :lgtm_strong:. </details>


---

Reviewed 3 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHqMHpm2F-u957nlWJv:335724814)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
To be honest I do not actually know, I had the same question. However we are in the process of contacting Erwin Coumans in [this Bullet issue](https://github.com/bulletphysics/bullet3/issues/34) and hopefully he will be able to give us feedback on why the simplex convex cast fails and also on if any of these algorithms are also applicable to concave geometries. Judging by the name only it'd make you think that both of these algorithms are only for convex geometries.

<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Very nice.  Thanks!  To be clear, judging by the name this will only work for convex geometries, correct?</details>


---

Review status: 4 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 615 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHprZfbhVvWbu08jeOr:-KHqdW8Ied6uIQzdZfK2:676635175) ([raw file](https://github.com/robotlocomotion/drake/blob/eaa05a61a2c9587d78dc60e1143b582334f2c2f2/drake/systems/plants/collision/bullet_model.cc#L615)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> examples?</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHqbuHy5Pr1WA_wVt1h:-1724659863)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> To be honest I do not actually know, I had the same question. However we are in the process of contacting Erwin Coumans in [this Bullet issue](https://github.com/bulletphysics/bullet3/issues/34) and hopefully he will be able to give us feedback on why the simplex convex cast fails and also on if any of these algorithms are also applicable to concave geometries. Judging by the name only it'd make you think that both of these algorithms are only for convex geometries.</details>


---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 624 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFg-LwkZGmxVspcmT:-KHqiOkB7cKyoI4nAP__:-993976015) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L624)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I think the comment is ok. I understand it as referring not to a particular convex object but to the class of objects that are convex. In any case, I am quoting here so my preference is to leave it as it is in the original. I'm open though</details>
In this case, "algorithm" should be "algorithms". Regardless, I'll mark this thread as resolved and leave the grammar fix to be at your discretion.

---

*[drake/systems/plants/collision/bullet_model.cc, line 626 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHkFoRlD4jWuhni0-Hp:-KHqoWv6WAQaHe06JaCU:178293334) ([raw file](https://github.com/robotlocomotion/drake/blob/5567d23181ddaecffcb1a088d5f56c074c233344/drake/systems/plants/collision/bullet_model.cc#L626)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I changed to "a very large". Is the "quite" the problem? I am trying to stress that this hard coded tolerance is really ridiculously large (1 cm. Imagine what would happen when simulating Valkyrie fingers!). My English might be failing though :-)</details>
The word "quite" sounded strange to me, though I'm not sure if its use in this context is actually wrong. The switch to "very" is good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHqosGWk-bqKX2nIERH:-1868698381)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@jwnimmer-tri would you mind being a platform reviewer for this PR?

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> :lgtm_strong: </details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHtql9JIHB26ruiROgy:-1626558801)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

jwnimmer-tri commented:
+@jwnimmer-tri  Sure.  I think you already have the required two LGTMs, but I'll take a brief read-through anyway.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> +@jwnimmer-tri would you mind being a platform reviewer for this PR?</details>


---

Reviewed 3 of 5 files at r2, 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 124 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtwA9Neg52KxfCwpJE:-KHtwA9Neg52KxfCwpJF:1620629699) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L124)):*
> ```objc
>       double collision_threshold) override;
> 
>   // Possible ray casting algorithms for Bullet.
> ```

Missing doxygen on public type.

---

*[drake/systems/plants/collision/bullet_model.h, line 125 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtw24nXJ5Bk61sc0Pn:-KHtw24nXJ5Bk61sc0Po:-392554147) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L125)):*
> ```objc
> 
>   // Possible ray casting algorithms for Bullet.
>   enum RayCastAlgorithm {
> ```

Unclear why this needs to be public; no public APIs use it.

---

*[drake/systems/plants/collision/bullet_model.h, line 162 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtwnY5Iu5Obi3tcCz6:-KHtwnY5Iu5Obi3tcCz7:-993501868) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L162)):*
> ```objc
>   // The Bullet algorithm to be used for ray casting.
>   // See notes in BulletModel::collisionRaycast.
>   RayCastAlgorithm ray_cast_algorithm_{kGjkConvexCast};
> ```

Unclear why this needs to be a member field (nevermind a non-const once), since it is never changed, nor can it be changed through any API.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHtvCUBwIzAJNrgFPeR:1888771935)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

amcastro-tri commented:
I thought so but I was unsure about who was the platform reviewer. That was my fault for not explicitly stating it. Who merges this then?

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> +@jwnimmer-tri  Sure.  I think you already have the required two LGTMs, but I'll take a brief read-through anyway.</details>


---

Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 124 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtwA9Neg52KxfCwpJE:-KHu5FoakDWXdUw8Mzew:738490302) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L124)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing doxygen on public type.</details>
Good!. Added doc. I wil definitely add more once we get a hold of Erwin Coumans (the creator of Bullet) in [this](https://github.com/bulletphysics/bullet3/issues/34) issue. 

---

*[drake/systems/plants/collision/bullet_model.h, line 125 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtw24nXJ5Bk61sc0Pn:-KHtyOLJ5Gsvl5Jp9JyN:806811558) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L125)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unclear why this needs to be public; no public APIs use it.</details>
Very good point. I was actually trying to expose this to the user. However I failed miserably, see issue #2361 that explains this with a specific example (actually this was code I wanted to submit here).

---

*[drake/systems/plants/collision/bullet_model.h, line 162 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtwnY5Iu5Obi3tcCz6:-KHu63QueSYdPwhJvoBs:1486469918) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unclear why this needs to be a member field (nevermind a non-const once), since it is never changed, nor can it be changed through any API.</details>
I know. See my previous reply and issue #2361 in an attempt to expose this option to the user. I would like to leave it here because I think it is cleaner and hopefully I will address #2361 soon to make it even cleaner. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHtxnN4JZApeqvwl_dh:24054688)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

sherm1 commented:
I'll merge it -- are you ready for it to go in?

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I thought so but I was unsure about who was the platform reviewer. That was my fault for not explicitly stating it. Who merges this then?</details>


---

Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHuDMDY8dhPM9GWRSRS:-1060681533)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewable says there are unresolved conversations and the branch is out of date w.r.t. master.

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I'll merge it -- are you ready for it to go in?</details>


---

Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHuDosfDbkfxhzdRG7V:1839218508)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 125 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtw24nXJ5Bk61sc0Pn:-KHuFDKrWTbwd99ZeIFz:404694606) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L125)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Very good point. I was actually trying to expose this to the user. However I failed miserably, see issue #2361 that explains this with a specific example (actually this was code I wanted to submit here).</details>
I still disagree that dead code should go into the tree.  If a future PR wants this, let the future PR commit it, not this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 125 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHtw24nXJ5Bk61sc0Pn:-KHuwuISSbaTo0h7VM83:-1119715311) ([raw file](https://github.com/robotlocomotion/drake/blob/b6cd436b93ed5e3f3436ee567662d2f46ec548e6/drake/systems/plants/collision/bullet_model.h#L125)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I still disagree that dead code should go into the tree.  If a future PR wants this, let the future PR commit it, not this one.</details>
It;'s not dead code! I did however made it private to the developers though. I agree the user should not see this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 637 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9ebPFQhp6JgkH36N:-KHw9ebPFQhp6JgkH36O:-719453825) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.cc#L637)):*
> ```cpp
>     // iterations, clip it to 32 for degenerate cases. See discussion about this
>     // here http://continuousphysics.com/Bullet/phpBB2/viewtopic.php?t=565
>     if (ray_cast_algorithm_ == kGjkConvexCast) {
> ```

Confusing if statement -- the when-false branch (omitting the flags change) is never possible to achieve.

---

*[drake/systems/plants/collision/bullet_model.h, line 159 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9jDwOBkvdB9UL_eb:-KHw9jDxuMdBWCg3aX5H:1402684972) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.h#L159)):*
> ```objc
>     // Implements Gino van den Bergens' paper "Ray Casting against General
>     // Convex Objects with Application to Continuous Collision Detection"
>     kSubSimplexConvexCast,
> ```

(Continuing the discussion here; reviewable didn't realize to move the comment thread over here.)

This value (SubSimplex) is never used anywhere.  It is dead code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@russtedrake the answer to your question:
>judging by the name this will only work for convex geometries, correct?

has arrived [here](https://github.com/bulletphysics/bullet3/issues/34). Quoting @erwincoumans:
> Q: Are these algorithms (GJK and sub sumplex) applicable to concave shapes as well?

> A: Yes, they can be extended to concave shapes, by using an acceleration structure, and doing a pairwise test for all potential candidates and taking the minimum. So although the elementary test only applies on convex shapes, the children in a concave shape are convex (convex triangles in a concave triangle mesh, or convex child shapes in a compound shape).



---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHwAEUMkeqnjNEAugq3:-1106531021)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 637 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9ebPFQhp6JgkH36N:-KHwEaD2cHDiSAVQIcty:610327451) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.cc#L637)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Confusing if statement -- the when-false branch (omitting the flags change) is never possible to achieve.</details>
Removed

---

*[drake/systems/plants/collision/bullet_model.h, line 159 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9jDwOBkvdB9UL_eb:-KHwEFvvr7w6RLqUI8Eo:673182817) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.h#L159)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> (Continuing the discussion here; reviewable didn't realize to move the comment thread over here.)
> 
> This value (SubSimplex) is never used anywhere.  It is dead code.</details>
You are right. I removed dead code and updated documentation so developers are aware of these options and to include recent conversation with Erwin Coumans (Bullet's developer). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 159 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9jDwOBkvdB9UL_eb:-KHwF2RzREhR7ZGfsPWB:79191735) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.h#L159)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> You are right. I removed dead code and updated documentation so developers are aware of these options and to include recent conversation with Erwin Coumans (Bullet's developer). </details>
Done.

(In retrospect, going back to read over the earlier thread, I see that could have been clearer about what I meant.  Sorry!  Sometimes I need to slow down before pushing 'Publish'.)

---

*[drake/systems/plants/collision/bullet_model.h, line 10 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHwFcG33hdGwtQQDGWW:-KHwFcG33hdGwtQQDGWX:843481339) ([raw file](https://github.com/robotlocomotion/drake/blob/b4956ff2fc22aaa6f1adbbe70967bea89d473b6b/drake/systems/plants/collision/bullet_model.h#L10)):*
> ```objc
> #include "BulletCollision/NarrowPhaseCollision/btGjkPairDetector.h"
> #include "BulletCollision/NarrowPhaseCollision/btPointCollector.h"
> #include "BulletCollision/NarrowPhaseCollision/btRaycastCallback.h"
> ```

I believe that this `#include`, the only remaining change to this file, is now unused by the code in the header?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 159 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHw9jDwOBkvdB9UL_eb:-KHwMgt2Gv1fJt_cnIay:-851142544) ([raw file](https://github.com/robotlocomotion/drake/blob/bac80e99a7ca27b5d3d2604dd416110e2cf0348a/drake/systems/plants/collision/bullet_model.h#L159)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Done.
> 
> (In retrospect, going back to read over the earlier thread, I see that could have been clearer about what I meant.  Sorry!  Sometimes I need to slow down before pushing 'Publish'.)</details>
No worries. Sometimes I need to slow down before doing `push` :-)

---

*[drake/systems/plants/collision/bullet_model.h, line 10 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHwFcG33hdGwtQQDGWW:-KHwMZ5HIuu1ujFyPcW4:-336023170) ([raw file](https://github.com/robotlocomotion/drake/blob/b4956ff2fc22aaa6f1adbbe70967bea89d473b6b/drake/systems/plants/collision/bullet_model.h#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I believe that this `#include`, the only remaining change to this file, is now unused by the code in the header?</details>
you mean I should move it to the cpp where it is used?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 10 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHwFcG33hdGwtQQDGWW:-KHwNXHFHjw8ggwiEwOd:1753504696) ([raw file](https://github.com/robotlocomotion/drake/blob/b4956ff2fc22aaa6f1adbbe70967bea89d473b6b/drake/systems/plants/collision/bullet_model.h#L10)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> you mean I should move it to the cpp where it is used?</details>
Right.

In general, the set of `#include` statements in the `.h` file should be exactly the set of includes that the code in the h file directly refers to.  The `#include` statements in the `.cc` file should be exactly the set of includes that the code in the cc file directly refers to.

In this case, looking at the entire PR, the single `#include` addition here stood out to me.  If nothing in btRaycastCallback is used by this header, the include statement should be removed.  (If the cc file then needs to add the include, so be it.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.h, line 10 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-KHwFcG33hdGwtQQDGWW:-KHx-5f2nxEHx9cytJXs:-682501828) ([raw file](https://github.com/robotlocomotion/drake/blob/b4956ff2fc22aaa6f1adbbe70967bea89d473b6b/drake/systems/plants/collision/bullet_model.h#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Right.
> 
> In general, the set of `#include` statements in the `.h` file should be exactly the set of includes that the code in the h file directly refers to.  The `#include` statements in the `.cc` file should be exactly the set of includes that the code in the cc file directly refers to.
> 
> In this case, looking at the entire PR, the single `#include` addition here stood out to me.  If nothing in btRaycastCallback is used by this header, the include statement should be removed.  (If the cc file then needs to add the include, so be it.)</details>
Good. Moved to .cc file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @russtedrake the answer to your question:
> >judging by the name this will only work for convex geometries, correct?
> 
> has arrived [here](https://github.com/bulletphysics/bullet3/issues/34). Quoting @erwincoumans:
> > Q: Are these algorithms (GJK and sub sumplex) applicable to concave shapes as well?
> 
> > A: Yes, they can be extended to concave shapes, by using an acceleration structure, and doing a pairwise test for all potential candidates and taking the minimum. So although the elementary test only applies on convex shapes, the children in a concave shape are convex (convex triangles in a concave triangle mesh, or convex child shapes in a compound shape).
> 
> </details>


---

Reviewed 2 of 2 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2354#-:-KHyUirGA8Ft-AJNSRlW:1417088721)*
<!-- Sent from Reviewable.io -->
